File: cli/src/main/java/io/microconfig/cli/CommandFactory.java
package io.microconfig.cli;

public class CommandFactory {
    public static Command command(String[] args) {
        if (args.length == 0) throw new CliException("No arguments", 1);

        switch (args[0]) {
            case "config":
                return new ConfigCommand(args);
            default:
                throw new CliException("Unsupported argument " + args[0], 2);
        }
    }
}

File: server-microctl/src/main/kotlin/io/microconfig/cli/commands/CommandFactory.kt
package io.microconfig.cli.commands

import io.microconfig.cli.CliException

val errorMessage = """
    Usage: microctl [command] [service] [flags]
    No command provided.
    Supported commands are [show, save, version, help]
    """.trimIndent()

fun command(args: Array<String>): Command {
    val cmd = System.getenv()["MC_CMD"]
    if (args.isEmpty() && cmd == null) throw CliException(errorMessage, 1)

    return when (val command = args.firstOrNull() ?: cmd) {
        "show" -> ShowCommand(args)
        "save" -> SaveCommand(args)
        "version" -> VersionCommand(args)
        "help" -> HelpCommand(args)
        else -> throw CliException("Unsupported argument $command", 2)
    }
}